source/RoslynAnalyzers/MessageContractAnalyzers.cs

```diff
@@ -0,0 +1,302 @@
+﻿using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System;
+using System.Diagnostics;
+using System.Threading;
+using static Octopus.RoslynAnalyzers.Descriptors;
+
+namespace Octopus.RoslynAnalyzers
+{
+    // Once we establish that a type is a "Message" (that is, it implements ICommand<>, IRequest<> or IResponse) then
+    // there's a whole variety of things that we want to assert against. We put them all in one analyzer to avoid the extra work that would
+    // be incurred if we had a dozen analyzers all going "is this a MessageType"?
+    // MessageContractAnalyzers use the ID range OCT3xxx
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public partial class MessageContractAnalyzers : DiagnosticAnalyzer
+    {
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+            EventTypesMustBeNamedCorrectly,
+            CommandTypesMustBeNamedCorrectly,
+            RequestTypesMustBeNamedCorrectly,
+            CommandTypesMustHaveCorrectlyNamedResponseTypes,
+            RequestTypesMustHaveCorrectlyNamedResponseTypes,
+            PropertiesOnMessageTypesMustBeMutable,
+            RequiredPropertiesOnMessageTypesMustNotBeNullable,
+            OptionalPropertiesOnMessageTypesMustBeNullable,
+            MessageTypesMustInstantiateCollections,
+            PropertiesOnMessageTypesMustHaveAtLeastOneValidationAttribute,
+            SpaceIdPropertiesOnMessageTypesMustBeOfTypeSpaceId,
+            IdPropertiesOnMessageTypesMustBeACaseInsensitiveStringTinyType,
+            MessageTypesMustHaveXmlDocComments,
+            ApiContractTypesMustLiveInTheAppropriateNamespace);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(CacheCommonTypes);
+            context.RegisterSyntaxNodeAction(CheckNode, SyntaxKind.ClassDeclaration, SyntaxKind.StructDeclaration);
+        }
+
+        SpecialTypeDeclarations? cachedTypeStorage; 
+
+#pragma warning disable RS1012 // context does not register any analyzer actions
```

Comment:
Full disclosure; I may well be overcooking it here and the optimization may not be necessary.

Other than using MSBuild ReportAnalyzer - which only provides a high level summary - I've not done any actual profiling on the code yet.

Over time as we write more analyzers I want dig into the performance/profiling side a bit more.
At that later point this optimization might get deleted (or justified), 

For now, what I'd like to do is merge this (unless someone objects), and get the analyzers working so they can start being helpful. Step 1 along a journey rather than final solution